### PR TITLE
moving a lot of validation for cluster create requests from

### PR DESCRIPTION
**Pull request by albertshau**
solve time to request time, as well as adding additional checks
for required provider fields.

### CODE REVIEW

**Comment by abaranau** [comment]
not mergeable

File: server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java [hunk 1]
```diff
@@ -502,4 +481,144 @@ public void changeExpireTime(String clusterId, Account account, long expireTime)
...
+
+    EntityStoreView entityStore = entityStoreService.getView(account);
+    // make sure the template exists
+    ClusterTemplate template = entityStore.getClusterTemplate(templateName);
+    if (template == null) {
+      throw new MissingEntityException("cluster template " + templateName + " does not exist.");
+    }
+
+    // check cluster size constraints
+    SizeConstraint sizeConstraint = template.getConstraints().getSizeConstraint();
```
Comment (abaranau) [inline server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java:521]: Can we encapsulate this within SizeContraint? E.g. 

```
sizeContraint.verify(numMachines);
```

Comment (albertshau) [inline server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java:521]: yep thats a good idea

File: server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java [hunk 2]
```diff
@@ -502,4 +481,144 @@ public void changeExpireTime(String clusterId, Account account, long expireTime)
...
+      throw new InvalidClusterException("Cluster size cannot be below " + minMachines + " nodes.");
+    }
+    if (numMachines > maxMachines) {
+      String errMsg = "Cluster size cannot exceed " + maxMachines;
+      errMsg += maxMachines == 1 ? " node." : " nodes.";
+      throw new InvalidClusterException(errMsg);
+    }
+    builder.setClusterTemplate(template);
+
+    setProvider(builder, account, template, createRequest, entityStore);
```
Comment (abaranau) [inline server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java]: may be cleaner:

```
Provider provider = getProviderAndValidate(account, template, createRequest, entityStore);
builder.setProvider(provider);
```

up to you ;)

Comment (abaranau) [inline server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java]: and others too

File: server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java [hunk 3]
```diff
@@ -502,4 +481,144 @@ public void changeExpireTime(String clusterId, Account account, long expireTime)
...
+  private void setCreateExpireTimes(Cluster.Builder builder, ClusterTemplate template,
+                                    ClusterCreateRequest createRequest) {
+    // Determine valid lease duration for the cluster.
+    // It has to be less than the initial lease duration set in template.
+    long requestedLease = createRequest.getInitialLeaseDuration();
+    long templateLease = template.getAdministration().getLeaseDuration().getInitial();
+    long leaseDuration;
+
+    // if it's -1, use the lease specified in the template
+    if (requestedLease == -1) {
```
Comment (abaranau) [inline server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java]: Similarly, does it make sense to incapsulate this in LeaseDuration returned from template? May be I'm wrong thought

File: server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java [hunk 4]
```diff
@@ -502,4 +481,144 @@ public void changeExpireTime(String clusterId, Account account, long expireTime)
...
+
+  private void setCreateExpireTimes(Cluster.Builder builder, ClusterTemplate template,
+                                    ClusterCreateRequest createRequest) {
+    // Determine valid lease duration for the cluster.
+    // It has to be less than the initial lease duration set in template.
+    long requestedLease = createRequest.getInitialLeaseDuration();
+    long templateLease = template.getAdministration().getLeaseDuration().getInitial();
+    long leaseDuration;
+
+    // if it's -1, use the lease specified in the template
```
Comment (abaranau) [inline server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java]: seems like this logic should be covered with unit-test somehow...

File: server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java (hunk 3 above)
Comment (albertshau) [inline server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java]: Yes it makes sense to move it to LeaseDuration.

File: server/src/main/java/com/continuuity/loom/spec/plugin/ProviderType.java [hunk 5]
```diff
@@ -27,4 +29,34 @@ public ProviderType(String name, String description, Map<ParameterType, Paramete
...
+      // see if this is an overridable admin field
+      FieldSchema fieldSchema = adminFields.get(field);
+      if (fieldSchema == null || !fieldSchema.getOverride()) {
+        // not an overridable admin field. check if its a user field
+        fieldSchema = userFields.get(field);
+      }
+
+      // if its not a user field or an overridable admin field, ignore it
+      if (fieldSchema != null) {
+        filtered.put(field, fieldVal);
```
Comment (abaranau) [inline server/src/main/java/com/continuuity/loom/spec/plugin/ProviderType.java:60]: should we log that we ignore some fields provided? Is it exception situation?

**Comment by abaranau** [comment]
nice code :) please make mergeable

File: server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java (hunk 4 above)
Comment (albertshau) [inline server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java]: it's tested indirectly through cluster create tests, but once I move it into LeaseDuration I can add some more direct tests.

File: server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java (hunk 2 above)
Comment (albertshau) [inline server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java]: makes sense, more re-usable that way too.

File: server/src/main/java/com/continuuity/loom/spec/plugin/ProviderType.java (hunk 5 above)
Comment (albertshau) [inline server/src/main/java/com/continuuity/loom/spec/plugin/ProviderType.java:60]: it is a sign that something unintended is probably happening. Will log it.

File: server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java (hunk 1 above)
Comment (abaranau) [inline server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java:521]: I see it became easy to test it :)

**Comment by abaranau** [comment]
LGTM

**Pull request merged**